[Chan Chang Yu, Linus] iP

linuschancs opened the pull request:
### BetaGo

> “The way to get started is to quit talking and begin doing.” – Walt Disney [Source](https://www.quotespedia.org/authors/w/walt-disney/the-way-to-get-started-is-to-quit-talking-and-begin-doing-walt-disney/)

BetaGo frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
5. And it is **FREE**!

Features:

- [x]  Managing tasks
- [ ]  Managing deadlines (_coming soon_)
- [ ]  Reminders (_coming soon_)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```ruby
public class Main {
    public static void main(String[] args) {
        new Duke().run()
    }
}
```

Commit:
Add Gradle support

linuschancs committed:
Implemented Greetings and commands will be echoed until the user enters 'bye' where the program will exit

linuschancs committed:
Added list functionality where users may store text and display the text they stored by entering 'list'

linuschancs committed:
Added new class for Tasks with class methods to mark tasks as done/undone, get description and get status icon

linuschancs committed:
Added Tasks and ability to mark tasks

linuschancs committed:
Added Deadline/Event/Todo classes

linuschancs committed:
Added tracking of Deadline/Event/Task and counter for lists

linuschancs committed:
Added a new TaskList class that makes use of ArrayLists to store the list of tasks.

linuschancs committed:
Added exception handling to certain cases such as when an InvalidCommand is provided

linuschancs committed:
Added testing using input/output redirection technique

linuschancs committed:
Added delete tasks functionality

linuschancs committed:
Abstract methods out into Duke class and called methods in Main

linuschancs committed:
Add save functionality to Duke.

Program will read from duke.txt file in data folder and load the task list accordingly.
If the file/folder does not exist, the program will create a folder/file.
Upon each change to the tasklist, mark/unmark/delete/add, the duke.txt file will be updated.

linuschancs committed:
Add Date and Time recognition into Deadline tasks

Upon adding a deadline task, the date/time needs to have the following format:
- yyyy-MM-dd
- dd-MMM-yyyy
- dd/MM/yyyy
- E, MMM dd yyyy

The date will then be converted to 'MMM d yyyy' accordingly

linuschancs committed:
Add Date and Time into deadline tasks

Users may enter dates in the following format: "yyyy-MM-dd, dd-MMM-yyyy, dd/MM/yyyy"
The program will then translate it to this format: "MMM d yyyy"

linuschancs committed:
Merge branch 'branch-Level-8'

linuschancs committed:
Fix Deadline bugs when converting from data file into list

linuschancs committed:
Extend use of OOP by asbtracting methods into different classes:
- Storage: Deals with loading tasks from the file and saving tasks in the file
- Ui: Deals with printing messages for the user
- Parser: Deals with making sense of the user command

linuschancs committed:
Divide classes into multiple packages

linuschancs committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip into add-gradle-support

linuschancs committed:
Added Gradle Support

linuschancs committed:
Add JUnit tests for creation of Deadline/Event/Todo tasks

linuschancs committed:
Edit code to follow coding standards

linuschancs committed:
Add JavaDoc comments to classes and methods

linuschancs committed:
Add Find functionality where users may enter a keyword and Duke will return a list of tasks whose description contains the keyword

linuschancs committed:
Merged branch-A-JavaDoc into master

linuschancs committed:
Merge branch-Level-9 into master

SilasTSL commented on src/main/java/betago/TaskList.java:
```diff
@@ -0,0 +1,317 @@
+     *
+     * @return Number of items in the ArrayList.
+     */
+    public int size() {
```
Should be named a verb, maybe getSize?

SilasTSL commented on src/main/java/betago/TaskList.java:
```diff
@@ -0,0 +1,317 @@
+     * @param str Delete command that the user provided.
+     * @throws InvalidCommandException If no task number is provided or task number is out of range.
+     */
+    public void deleteItems(String str) throws InvalidCommandException {
```
Should be deleteItem because it only deletes one task?

SilasTSL commented on src/main/java/betago/TaskList.java:
```diff
@@ -0,0 +1,317 @@
+     * @param str Mark or Unmark command that the user provided.
+     * @throws InvalidCommandException If there are no Tasks in the list.
+     */
+    public void markUnmarkItems(String str) throws InvalidCommandException {
```
Should be markUnmarkItem, since it only marks/unmarks a single Task?

SilasTSL commented on src/main/java/betago/tasks/Task.java:
```diff
@@ -0,0 +1,61 @@
+        return "";
+    }
+
+    public boolean containKeyword(String keyword) {
```
Should be containsKeyword?

SilasTSL reviewed (commented):
Overall LGTM, nice job with OOP!

linuschancs committed:
Fixed Checkstyle errors and added more JavaDocs

linuschancs committed:
Fixed some print statement bugs

jia-han commented on src/main/java/betago/Parser.java:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Reads input from the user and calls the corresponding methods.
+     */
+    public void readCommands() {
```
Perhaps the use of different command tasks would be better here to encapsulate the try catch logic?

jia-han commented on src/main/java/betago/TaskList.java:
```diff
@@ -0,0 +1,317 @@
+        }
+        System.out.print("Here are the tasks in your list:\n");
+        for (int i = 0; i < this.list.size(); i++) {
+            System.out.print(i+1);
```
There should be spaces surrounding the operator character?

jia-han commented on src/main/java/betago/TaskList.java:
```diff
@@ -0,0 +1,317 @@
+                    System.out.println(this.list.get(marker-1).toString() + "\n");
+                }
+            }
+            catch (NumberFormatException ex){
```
The catch statement should not be on a new line?

jia-han reviewed (commented):
Overall, the code is easily understandable. However, the code in both Parser and Tasklist class could be abstracted out into functions and classes instead of being repeated.

linuschancs committed:
Implemented a GUI using JavaFX and refactored classes.

Commands is now an interface and the various types of commands are separate classes that implements this interface.
Exceptions have also been reduced to a single DukeException.
UI class now handles the printing of certain errors and greet statement.

linuschancs committed:
Made minor changes to some output statements to improve readability

linuschancs committed:
Added Javadocs comments and headers

linuschancs committed:
Added some styling to DialogBox and made sure full output can be viewed.

linuschancs committed:
Add storage of Date and Time to Event Task

Users could previously store any String into the /at field of Event task.
This feature limits users to store a proper Date into the Event task,
and converts this date into a standardised date format for better readability.

Let's only allow users to enter their dates for the Event task in the following format:
"yyyy-MM-dd", "dd-MMM-yyyy", "dd/MM/yyyy".
The date will then be stored as "MMM d yyyy".

linuschancs committed:
Edit Parser to use String input directly rather than redundant Scanner

linuschancs committed:
Add use of Assertions into TaskList load deadline and event methods.

Assertions used to document important assumptions that should hold at that point
so that runtime can verify them. Helps early detection of errors.

linuschancs committed:
Improve code quality by critically examining and refactoring code.

Code quality affects not only run-time effeciency, security, and robustness,
but also improves reader understandability.

Let's,
- Avoid long methods (below 30 LOC)
- Avoid deep nesting and arrowhead style code
- Add default branches for case statements
- Minimise code duplication
- Use guard clauses

linuschancs committed:
Merge pull request #1 from linuschancs/branch-A-CodeQuality

Improve code quality by critically examining and refactoring code.

linuschancs committed:
Merged branch-A-Assertions with master

linuschancs committed:
Merge pull request #2 from linuschancs/branch-A-Assertions

Merge Branch-A-Assertions with master

linuschancs committed:
Add functionality of specifying specific file to load data from.

Current data is always pulled and saved from duke.txt file.

Allowing users to specify the file they want to load and save data in allows
them to personalise multiple tasklists and adds more personalisation.

Let's,
- Make default file load and save into duke.txt file
- Allow users to call a new load command to specify the file they want to load
- After successful load, auto-save will occur into the new file instead of duke.txt file

However, current implementation clears tasklist when loading data even when
file is not able to load due to invalid entries which may cause data to be lost.

linuschancs committed:
Merge pull request #3 from linuschancs/branch-C-FlexibleDataSource

Add functionality of specifying specific file to load data from.

linuschancs committed:
Add Continuous Integration using GitHub Actions

linuschancs committed:
Change display pictures to add in more personality to program

linuschancs committed:
Add title BetaGo to GUI window

linuschancs committed:
Add representative screenshot to docs folder

linuschancs committed:
Add user guide into docs/README.md

linuschancs committed:
Merge pull request #5 from linuschancs/branch-A-UserGuide

Add user guide into docs/README.md

linuschancs committed:
Update User Guide

linuschancs committed:
Merge branch 'master' of https://github.com/linuschancs/ip

linuschancs committed:
Fix .jar file error with Gradle

linuschancs committed:
Update README.md

linuschancs committed:
Update README.md

linuschancs committed:
Update README.md

linuschancs committed:
Update README.md

linuschancs committed:
Update README.md

linuschancs committed:
Fix deadline and event commands and update README to reflect new command formats

linuschancs committed:
Merge branch 'master' of https://github.com/linuschancs/ip

linuschancs committed:
Update README.md
